lxd/auth/driver_rbac.go

```diff
@@ -127,199 +108,251 @@ func (r *rbac) load() error {
+	}
 
-			err = json.NewDecoder(resp.Body).Decode(&status)
-			_ = resp.Body.Close()
-			if err != nil {
-				logger.Errorf("Failed to parse RBAC response, re-trying: %v", err)
-				time.Sleep(5 * time.Second)
-				continue
-			}
+	permission, err := r.relationToPermission(object, entitlement)
+	if err != nil {
+		return err
+	}
 
-			r.lastChange = status.LastChange
-			logger.Debugf("RBAC change detected, flushing cache")
-			r.flushCache()
-		}
-	}()
-}
+	projectName := object.Project()
+	if !shared.ValueInSlice(permission, permissions[projectName]) {
+		return api.StatusErrorf(http.StatusForbidden, "User %q does not have permission %q on project %q", username, permission, projectName)
+	}
 
-// StopStatusCheck stops the periodic status checker.
-func (r *rbac) StopStatusCheck() {
-	r.ctxCancel()
+	return nil
 }
 
-// syncProjects updates the list of projects in RBAC.
-func (r *rbac) syncProjects() error {
-	if r.projectsGetFunc == nil {
-		return fmt.Errorf("ProjectsFunc isn't configured yet, cannot sync")
+// GetPermissionChecker syncs the users permissions with the RBAC server, then in the returned PermissionChecker maps the
+// given Object and Entitlement to an RBAC permission and checks this against the users permissions.
+func (r *rbac) GetPermissionChecker(ctx context.Context, req *http.Request, entitlement Entitlement, objectType ObjectType) (PermissionChecker, error) {
+	allowFunc := func(b bool) func(Object) bool {
+		return func(Object) bool {
+			return b
+		}
 	}
 
-	resources := []rbacResource{}
-	resourcesMap := map[string]string{}
-
-	// Get all projects
-	projects, err := r.projectsGetFunc()
+	details, err := r.requestDetails(req)
 	if err != nil {
-		return err
+		return nil, api.StatusErrorf(http.StatusForbidden, "Failed to extract request details: %v", err)
 	}
 
-	// Convert to RBAC format
-	for id, name := range projects {
-		resources = append(resources, rbacResource{
-			Name:       name,
-			Identifier: strconv.FormatInt(id, 10),
-		})
+	if details.isInternalOrUnix() {
+		return allowFunc(true), nil
+	}
 
-		resourcesMap[name] = strconv.FormatInt(id, 10)
+	// Use the TLS driver if the user authenticated with TLS.
+	if details.authenticationProtocol() == api.AuthenticationMethodTLS {
+		return r.tls.GetPermissionChecker(ctx, req, entitlement, objectType)
 	}
 
-	// Update RBAC
-	err = r.postResources(resources, nil, true)
-	if err != nil {
-		return err
+	r.permissionsLock.Lock()
+	defer r.permissionsLock.Unlock()
+
+	username := details.username()
+	permissions, ok := r.permissions[username]
+	if !ok {
+		err := r.syncPermissions(ctx, username)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to sync user permissions with RBAC server: %w", err)
+		}
+
+		permissions, ok = r.permissions[username]
+		if !ok {
+			return nil, errUnknownUser
+		}
 	}
 
-	// Update project map
-	r.resourcesLock.Lock()
-	r.resources = resourcesMap
-	r.resourcesLock.Unlock()
+	if shared.ValueInSlice(PermissionAdmin, permissions[""]) {
```

Comment:
@markylaing I dont understand what this part is doing, will it always be false?

Reply:
The `permissions` map here is a map of project name to slice of permissions. However, if the user is an admin, this permission is stored in the map with the project name as an empty string. This is how it is currently done, but I can change it to something like:
```golang
type rbacPermission struct {
    admin bool
    projects map[string][]Permission
}
```

The permission cache defined on `rbac` will then be a `map[string]rbacPermission` where the keys are usernames.

How does that sound?

Reply:
yeah i think that would be clearer. Thanks